apis/database/v1alpha1/redshift_types.go

```diff
@@ -0,0 +1,713 @@
+
+	// ClusterStatus is the current state of the cluster.
+	ClusterStatus string `json:"clusterStatus,omitempty"`
+
+	// ClusterSubnetGroupName is the name of a cluster subnet group to be associated with this cluster.
+	// If this parameter is not provided the resulting cluster will be deployed
+	// outside virtual private cloud (VPC).
+	ClusterSubnetGroupName string `json:"clusterSubnetGroupName,omitempty"`
+
+	// Describes the status of a cluster while it is in the process of resizing
+	// with an incremental resize.
+	DataTransferProgress DataTransferProgress `json:"dataTransferProgress,omitempty"`
+
+	// Describes a group of DeferredMaintenanceWindow objects.
+	DeferredMaintenanceWindows []DeferredMaintenanceWindow `json:"deferredMaintenanceWindows,omitempty"`
+
+	// The status of the elastic IP (EIP) address.
+	ElasticIPStatus ElasticIPStatus `json:"elasticIPStatus,omitempty"`
+
+	// EnhancedVPCRouting specifies whether to create the cluster with enhanced VPC
+	// routing enabled. To create a cluster that uses enhanced VPC routing, the
+	// cluster must be in a VPC. For more information, see Enhanced VPC Routing
+	// (https://docs.aws.amazon.com/redshift/latest/mgmt/enhanced-vpc-routing.html)
+	// in the Amazon Redshift Cluster Management Guide.
+	EnhancedVPCRouting bool `json:"enhancedVPCRouting,omitempty"`
+
+	// The number of nodes that you can resize the cluster to with the elastic resize
+	// method.
+	ElasticResizeNumberOfNodeOptions string `json:"elasticResizeNumberOfNodeOptions,omitempty"`
+
+	// Endpoint specifies the connection endpoint.
+	Endpoint Endpoint `json:"endpoint,omitempty"`
+
+	// The date and time when the next snapshot is expected to be taken for clusters
+	// with a valid snapshot schedule and backups enabled.
+	ExpectedNextSnapshotScheduleTime *metav1.Time `json:"expectedNextSnapshotScheduleTime,omitempty"`
+
+	// The status of next expected snapshot for clusters having a valid snapshot
+	// schedule and backups enabled. Possible values are the following:
+	//
+	//    * OnTrack - The next snapshot is expected to be taken on time.
+	//
+	//    * Pending - The next snapshot is pending to be taken.
+	ExpectedNextSnapshotScheduleTimeStatus string `json:"expectedNextSnapshotScheduleTimeStatus,omitempty"`
+
+	// A value that reports whether the Amazon Redshift cluster has finished applying
+	// any hardware security module (HSM) settings changes specified in a modify
+	// cluster command.
+	//
+	// Values: active, applying
+	HSMStatus HSMStatus `json:"hsmStatus,omitempty"`
+
+	// IAMRoles is a list of AWS Identity and Access Management (IAM) roles that can be used
+	// by the cluster to access other AWS services. You must supply the IAM roles
+	// in their Amazon Resource Name (ARN) format. You can supply up to 10 IAM roles
+	// in a single request.
+	// A cluster can have up to 10 IAM roles associated with it at any time.
+	IAMRoles []ClusterIAMRole `json:"iamRoles,omitempty"`
```

Comment:
This and a few other fields that are in the `spec` appear to be duplicated in the `status`. They should be sufficient in the `spec` where they can be late initialized

Reply:
I have removed `IAMRoles`, `EnhancedVPCRouting`, `KMSKeyID`, `PreferredMaintenanceWindow`, and,  `SnapshotScheduleIdentifier`.  I decided to keep some other fields that are also present in Parameters because AWS creates and sets a default for them if none is provided. So the idea was to be able to reflect those defaults. Let me know if I should go ahead and remove those other fields also(namely, `ClusterSubnetGroupName`, `ClusterSecurityGroups`, and `VPCSecurityGroups`).

Reply:
@arush-sal typically we would want to populate the spec with the defaults for these values that AWS sets if no values are provided by the user